### PR TITLE
Ipv6：icmpv6、toa、xmit

### PR DESCRIPTION
**Pull request by AdamYYan**

### CODE REVIEW

File: include/ipv6.h [hunk 1]
```diff
@@ -68,6 +73,11 @@ struct inet6_protocol {
     unsigned int    flags;
 };
 
+int ip6_skip_exthdr(const struct rte_mbuf *imbuf, int start,
+                    __u8 *nexthdrp);
```
Comment (mscbg) [inline include/ipv6.h]: function int ip6_skip_exthdr(const struct rte_mbuf *mbuf, int start, uint8_t *nexthdrp); should be deleted

File: src/ipv6/ipv6.c [hunk 2]
```diff
@@ -419,7 +420,8 @@ static int ip6_forward(struct rte_mbuf *mbuf)
         goto error;
 
     if (addrtype & IPV6_ADDR_LINKLOCAL) {
-        //icmpv6_send(mbuf, ICMPV6_DEST_UNREACH, ICMPV6_NOT_NEIGHBOUR, 0);
+        // don't exist ICMP6_NOT_NEIGHBOUR in netinet/icmp6.h, use ICMP6_DST_UNREACH_NOROUTE
+        icmp6_send(mbuf, ICMP6_DST_UNREACH, ICMP6_DST_UNREACH_NOROUTE, 0);
```
Comment (mscbg) [inline src/ipv6/ipv6.c]: icmp报文是否要匹配。

File: src/ipvs/ip_vs_proto_tcp.c [hunk 3]
```diff
@@ -690,7 +713,7 @@ static int tcp_fnat_out_handler(struct dp_vs_proto *proto,
     th->dest    = conn->cport;
 
     if (th->syn && th->ack)
-        tcp_out_adjust_mss(th);
+        tcp_out_adjust_mss(AF_INET, th);
```
Comment (ywc689) [inline src/ipvs/ip_vs_proto_tcp.c]: `tcp_out_adjust_mss(AF_INET, th);` -> `tcp_out_adjust_mss(af, th);`

File: src/ipv6/icmp6.c [hunk 4]
```diff
@@ -50,37 +67,176 @@ static inline void icmp6_send_csum(struct ip6_hdr *shdr, struct icmp6_hdr *ich)
...
+    int ptr = sizeof(ip6_hdr);
+    __u8 nexthdr = ip6_hdr(imbuf)->ip6_nxt;
+
+    if (mbuf_may_pull(imbuf, ptr) != 0) {
+        return 1;
+    }
+
+    ptr = ip6_skip_exthdr(imbuf, ptr, &nexthdr);
+    if (ptr < 0)
+        return 0;
```
[the remark this answers is not in the document]
Comment (AdamYYan) [reply src/ipv6/icmp6.c]: 参照内核实现，是return 0

File: src/ipv6/icmp6.c [hunk 5]
```diff
@@ -50,37 +67,176 @@ static inline void icmp6_send_csum(struct ip6_hdr *shdr, struct icmp6_hdr *ich)
...
+     */
+    if ((addr_type == IPV6_ADDR_ANY) || (addr_type & IPV6_ADDR_MULTICAST)) {
+        RTE_LOG(DEBUG, ICMP6, "icmpv6_send: addr_any/mcast source\n");
+        return;
+    }
 
-    if (mbuf_may_pull(mbuf, mbuf->pkt_len) != 0)
-        goto drop;
+    /*
+     *  Never answer to a ICMP packet.
```
[the remark this answers is not in the document]
Comment (AdamYYan) [reply src/ipv6/icmp6.c]: ok。这个comment有问题，已修改成Never answer to a ICMP packet except type of ICMP6_INFOMSG_MASK.

File: src/ipv6/icmp6.c [hunk 6]
```diff
@@ -50,37 +67,176 @@ static inline void icmp6_send_csum(struct ip6_hdr *shdr, struct icmp6_hdr *ich)
...
-    assert(iph);
+    addr_type = ipv6_addr_type(&iph->ip6_dst);
 
-    if (mbuf_may_pull(mbuf, sizeof(struct icmp6_hdr)) != 0)
-        goto drop;
+    if (addr_type & IPV6_ADDR_MULTICAST ||  etype != ETH_PKT_HOST) {
+        if (type != ICMP6_PACKET_TOO_BIG &&
+            !(type == ICMP6_PARAM_PROB &&
+              code == ICMP6_PARAMPROB_OPTION &&
+              (icmp6_opt_unrec(imbuf, info)))) {
```
[the remark this answers is not in the document]
Comment (AdamYYan) [reply src/ipv6/icmp6.c]: /*
     * when the original ipv6 dst is l2/l3 mcast, just deal ICMP6_PACKET_TOO_BIG and 
     * ICMP6_PARAM_PROB's unrecognize IPv6 option 
     */

File: src/ipv6/icmp6.c [hunk 7]
```diff
@@ -50,37 +67,176 @@ static inline void icmp6_send_csum(struct ip6_hdr *shdr, struct icmp6_hdr *ich)
...
-    /* ndisc test*/
-    if ((ich->icmp6_type == ND_NEIGHBOR_SOLICIT) ||
-       (ich->icmp6_type == ND_NEIGHBOR_ADVERT)) {
-        return ndisc_rcv(mbuf, netif_port_get(mbuf->port));
+    fl6.fl6_proto = IPPROTO_ICMPV6;
+    fl6.fl6_oif = netif_port_get(imbuf->port);
+    if (saddr) {
+        shdr.ip6_src = fl6.fl6_saddr = *saddr;
+    }
+    else {
```
[the remark this answers is not in the document]
Comment (AdamYYan) [reply src/ipv6/icmp6.c]: 好的

File: src/ipvs/ip_vs_proto_tcp.c (hunk 3 above)
Comment (AdamYYan) [reply src/ipvs/ip_vs_proto_tcp.c]: ok，我之前以为这个tcp_fnat_out_handler功能会分为v4/v6两个函数，预先写了AF_INET。 rebase的时候，没有太注意。

File: src/ipvs/ip_vs_core.c [hunk 8]
```diff
@@ -583,6 +709,9 @@ static int __dp_vs_in_icmp(struct rte_mbuf *mbuf, int *related)
         return INET_DROP;
     }
 
+    // re-fetch IP header and Icmp address
+    iph = ip4_hdr(mbuf);
+    ich = (struct icmphdr*)((void*)iph + ip4_hdrlen(mbuf));
```
[the remark this answers is not in the document]
Comment (AdamYYan) [reply src/ipvs/ip_vs_core.c]: base on the implement of  mbuf_may_pull( above this code line) now， the re-fetch IP is Superfluous. i think about use iph safely. so add this code.  of cause ip4_hdr(mbuf) is effective

File: src/ipvs/ip_vs_xmit.c [hunk 9]
```diff
@@ -223,9 +283,33 @@ static void dp_vs_conn_cache_rt(struct dp_vs_conn *conn, struct route_entry *rt,
     }
 }
 
-int dp_vs_xmit_fnat(struct dp_vs_proto *proto,
-                        struct dp_vs_conn *conn,
-                        struct rte_mbuf *mbuf)
+static void dp_vs_conn_cache_rt6(struct dp_vs_conn *conn, struct route6 *rt, bool in)
+{
+    if ((in && conn->in_dev && ipv6_addr_any(&conn->in_nexthop.in6)) ||
+        (!in && conn->out_dev && ipv6_addr_any(&conn->out_nexthop.in6)))
```
[the remark this answers is not in the document]
Comment (AdamYYan) [reply src/ipvs/ip_vs_xmit.c]: ok

File: src/ipvs/ip_vs_xmit.c [hunk 10]
```diff
@@ -16,26 +16,47 @@
...
+#include "icmp6.h"
 #include "neigh.h"
 #include "ipvs/xmit.h"
 #include "parser/parser.h"
 
 static bool fast_xmit_close = false;
 static bool xmit_ttl = false;
 
-static int dp_vs_fast_xmit_fnat(struct dp_vs_proto *proto,
+static int dp_vs_fast_xmit_fnat(int af,
```
[the remark this answers is not in the document]
Comment (AdamYYan) [reply src/ipvs/ip_vs_xmit.c]: ok

File: src/ipv6/ipv6.c (hunk 2 above)
Comment (AdamYYan) [reply src/ipv6/ipv6.c]: 嗯，是一个问题。

File: include/ipv6.h (hunk 1 above)
Comment (AdamYYan) [reply include/ipv6.h]: ok

**Pull request merged**